stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/viewmodel/ChannelListViewModel.kt

```diff
@@ -0,0 +1,130 @@
+
+    private val paginationStateMerger = MediatorLiveData<PaginationState>()
+    public val paginationState: LiveData<PaginationState> = Transformations.distinctUntilChanged(paginationStateMerger)
+
+    init {
+        stateMerger.value = State.Loading
+        chatDomain.useCases.queryChannels(filter, sort, limit).enqueue { queryChannelsControllerResult ->
+            if (queryChannelsControllerResult.isSuccess) {
+                val queryChannelsController = queryChannelsControllerResult.data()
+                stateMerger.addSource(
+                    map(queryChannelsController.channelsState) { channelState ->
+                        when (channelState) {
+                            is QueryChannelsController.ChannelsState.NoQueryActive,
+                            is QueryChannelsController.ChannelsState.Loading -> State.Loading
+                            is QueryChannelsController.ChannelsState.OfflineNoResults -> State.NoChannelsAvailable
+                            is QueryChannelsController.ChannelsState.Result ->
+                                State.Result(channelState.channels.filterNot { it.hidden == true || it.isDraft })
+                        }
+                    }
+                ) { state -> stateMerger.value = state }
+                paginationStateMerger.addSource(queryChannelsController.loadingMore) { loadingMore ->
+                    setPaginationState { copy(loadingMore = loadingMore) }
+                }
+                paginationStateMerger.addSource(queryChannelsController.endOfChannels) { endOfChannels ->
+                    setPaginationState { copy(endOfChannels = endOfChannels) }
+                }
+            } else {
+                stateMerger.postValue(State.NoChannelsAvailable)
+            }
+        }
+    }
+
+    public fun onAction(action: Action) {
+        when (action) {
+            is Action.ReachedEndOfList -> requestMoreChannels()
+            is Action.LogoutClicked -> {
+                ChatClient.instance().disconnect()
+                stateMerger.postValue(State.NavigateToLoginScreen)
+            }
+        }.exhaustive
+    }
+
+    public fun leaveChannel(channel: Channel) {
+        chatDomain.useCases.leaveChannel(channel.cid).enqueue()
+    }
+
+    public fun deleteChannel(channel: Channel) {
+        chatDomain.useCases.deleteChannel(channel.cid).enqueue()
+    }
+
+    public fun hideChannel(channel: Channel) {
+        chatDomain.useCases.hideChannel(channel.cid, true).enqueue()
+    }
+
+    public fun markAllRead() {
+        chatDomain.useCases.markAllRead().enqueue()
+    }
+
+    private fun requestMoreChannels() {
+        chatDomain.useCases.queryChannelsLoadMore(filter, sort).enqueue()
+    }
+
+    private fun setPaginationState(reducer: PaginationState.() -> PaginationState) {
+        paginationStateMerger.value = reducer(paginationStateMerger.value ?: PaginationState())
+    }
+
+    public sealed class State {
```

Comment:
I'd suggest making `State` a data class, and keeping the view model design consistent with other view models in the `ui` module, eg. `SearchViewModel` or `MentionListViewModel`.

Reply:
I think we should do it everywhere we have interactions with state & events so we don't lose values when we transition from one state to another. Is it within the scope of this PR, though?

Reply:
agree! let's do it in a separate PR. I would like to merge #1379 before that